core/corehttp/gateway_handler.go

```diff
@@ -161,38 +175,97 @@ func (i *gatewayHandler) getOrHeadHandler(w http.ResponseWriter, r *http.Request
 	}
 
 	// Resolve path to the final DAG node for the ETag
-	resolvedPath, err := i.api.ResolvePath(r.Context(), parsedPath)
-	switch err {
-	case nil:
-	case coreiface.ErrOffline:
-		webError(w, "btfs resolve -r "+escapedURLPath, err, http.StatusServiceUnavailable)
-		return
-	default:
-		webError(w, "btfs resolve -r "+escapedURLPath, err, http.StatusNotFound)
-		return
+	var (
+		resolvedPath              ipath.Resolved
+		rootPath                  string
+		escapedRootPath           string
+		resolvedRootPath          ipath.Resolved
+		dr                        files.Node
+		err                       error
+		isReedSolomonSubdirOrFile bool
+	)
+	top, err := i.isTopLevelEntryPath(r, parsedPath.String())
+	if err != nil {
+		switch err {
+		case nil:
+		case coreiface.ErrOffline:
+			webError(w, "btfs resolve -r "+escapedURLPath, err, http.StatusServiceUnavailable)
+			return
+		default:
+			webError(w, "btfs resolve -r "+escapedURLPath, err, http.StatusNotFound)
+			return
+		}
 	}
+	if !top {
+		// Check if the `parsedPath` is part of Reed-Solomon enconded directory object.
+		// If yes, put a new entry to the Reed-Solomon directory cache if not exist and
+		// get the files.Node of the `parsedPath`.
+		k := strings.SplitN(parsedPath.String(), "/", 4)[2]
```

Comment:
Add some slice bound checking in case of malformed string?

Reply:
Line 174 has 	if err := parsedPath.IsValid(); err != nil { 
for the parsedPath. So skipped checking.